### ISSUE TITLE
INSTALL: Installing Firedrake on a cluster

### ISSUE THREAD

**Issue by HenryCarr98** [firedrakeproject/firedrake#3949]
**Describe the error**
Attempting to run python3 firedrake-install  on a cluster returns 

```
FileNotFoundError: [Errno 2] No such file or directory: 'apt-get'
```

**Steps to Reproduce**

run python3 firedrake-install  on a cluster with no sudo rights

ie: The exact command used for installation/update  '...'

**Expected behavior**

Firedrake should install

**Error message**

```bash
python3 firedrake-install 
/rds/user/username/hpc-work/HPCSSUP-80609/firedrake-install:27: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
  from pkg_resources.extern.packaging.version import Version, InvalidVersion
Running firedrake-install
Traceback (most recent call last):
  File "/rds/user/username/hpc-work/HPCSSUP-80609/firedrake-install", line 1623, in <module>
    check_call(["apt-get", "--version"])
  File "/rds/user/username/hpc-work/HPCSSUP-80609/firedrake-install", line 680, in check_call
    log.debug(subprocess.check_output(arguments, stderr=subprocess.STDOUT, env=os.environ).decode())
  File "/home/username/rds/hpc-work/miniconda3/envs/Firedrake/lib/python3.10/subprocess.py", line 420, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/home/username/rds/hpc-work/miniconda3/envs/Firedrake/lib/python3.10/subprocess.py", line 501, in run
    with Popen(*popenargs, **kwargs) as process:
  File "/home/username/rds/hpc-work/miniconda3/envs/Firedrake/lib/python3.10/subprocess.py", line 966, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/home/username/rds/hpc-work/miniconda3/envs/Firedrake/lib/python3.10/subprocess.py", line 1842, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: 'apt-get'
```

**Environment:**
 - OS: RHEL8 Linux 
 - Python version: 3.10

**Additional context**

I would please like to know how to install this without using sudo or apt-get. Thanks.

**Comment by connorjward** [comment]
As per the documentation (https://www.firedrakeproject.org/download.html#system-requirements) you need to pass `--no-package-manager` to the installation. Note that it is then your responsibility to ensure all the correct dependencies are installed in advance.

**Comment by HenryCarr98** [comment]
Hi, thank you for your message. Indeed, I found that option but wanted to make sure it was correct. Is it normal that this process takes over an hour?

**Comment by HenryCarr98** [comment]
Not a problem, thank you for your help! And happy new year!!

**Comment by HenryCarr98** [comment]
Hi, I just wanted to update this item and ask if there's a simpler way to do this without sudo? And apt-get? Can I not manually disable it but let it build with the dependencies? Ie. Avoid having to use "--no-package-manager"? Speaking to my colleagues we'd be really pleased if e.g. a conda install became available! Thanks again.

**Comment by dham** [comment]
I think I don't understand what you are asking for. Firedrake has various dependencies, which are listed at https://www.firedrakeproject.org/download.html#system-requirements. If you have those dependencies installed, then you simply run the install script and pass `--no-package-manager` and Firedrake goes ahead and installs. I don't think I understand why you think that is not simple.

Running the default configuration will (on apt-based Linux systems) attempt to install the dependencies for you. This is what most users on workstations want, which is why it's the default. Apt requires sudo which means that it's not a suitable install mechanism on a cluster. `--no-package-manager` is provided for that case.

In other words, either you have to install dependencies, or you have to have us do it. I don't really see that there is a third option.

We don't have plans to support Anaconda. Anaconda inserts a whole compilation tool chain into the environment which conflicts with other tools. That's very difficult for Firedrake, which depends on runtime just-in-time compilation, to accommodate.

**Comment by HenryCarr98** [comment]
Right, and my question to you is how are users **without sudo access** supposed to install this? Fair enough, the dependencies are linked in that page but the average first time user is not going to understand this. My question is: what other install pathways are available that will suit less experienced users? I.e., conda, apptainer/docker images etc.

**Comment by dham** [comment]
We provide docker containers and singularity/apptainer instructions: https://github.com/firedrakeproject/firedrake/wiki/singularity

If you want to natively install on a cluster then you need to have the dependencies. That's hardly unique to Firedrake. None of the dependencies are particularly esoteric, and I would usually expect them to already be on a cluster or, at the very worst, be things that you can ask the cluster administrators to build for you. If all else fails, you can build them yourself from source. Once again, this is hardly a problem unique to Firedrake.

We work very hard to enable simple workstation installs for inexperienced users. When it comes to clusters, the situation is inherently more complicated. We don't have a magic wand that makes building software on clusters less fiddly than it is. This is an unfortunate fact of life that, so far, 30-odd years of cluster administration tools haven't overcome.

**Comment by HenryCarr98** [comment]
I see, thank you for providing the link. I'll mark this ticket as closed in this instance seeing as the answer is to use `--no-package-manager`. Thank you for your time.
